tests/unittest_nodes_lineno.py

```diff
@@ -0,0 +1,1185 @@
+
+        c2 = ast_nodes[1]
+        assert isinstance(c2, nodes.Const)
+        assert (c2.lineno, c2.col_offset) == (2, 0)
+        assert (c2.end_lineno, c2.end_col_offset) == (2, 7)
+
+        c3 = ast_nodes[2]
+        assert isinstance(c3, nodes.Const)
+        assert (c3.lineno, c3.col_offset) == (3, 0)
+        assert (c3.end_lineno, c3.end_col_offset) == (3, 4)
+
+        c4 = ast_nodes[3]
+        assert isinstance(c4, nodes.Const)
+        assert (c4.lineno, c4.col_offset) == (4, 0)
+        assert (c4.end_lineno, c4.end_col_offset) == (4, 4)
+
+        c5 = ast_nodes[4]
+        assert isinstance(c5, nodes.Const)
+        assert (c5.lineno, c5.col_offset) == (5, 0)
+        assert (c5.end_lineno, c5.end_col_offset) == (5, 5)
+
+        c6 = ast_nodes[5]
+        assert isinstance(c6, nodes.Const)
+        assert (c6.lineno, c6.col_offset) == (6, 0)
+        assert (c6.end_lineno, c6.end_col_offset) == (6, 3)
+
+    @staticmethod
+    def test_end_lineno_function() -> None:
+        """FunctionDef, AsyncFunctionDef, Decorators, Lambda, Arguments."""
+        code = textwrap.dedent(
+            """
+        def func(  #@
+            a: int = 0, /,
+            var: int = 1, *args: Any,
+            keyword: int = 2, **kwargs: Any
+        ) -> None:
+            pass
+
+        @decorator1
+        @decorator2
+        async def func():  #@
+            pass
+
+        lambda x: 2  #@
+        """
+        ).strip()
+        ast_nodes = builder.extract_node(code)
+        assert isinstance(ast_nodes, list) and len(ast_nodes) == 3
+
+        # fmt: off
+        f1 = ast_nodes[0]
+        assert isinstance(f1, nodes.FunctionDef)
+        assert isinstance(f1.args, nodes.Arguments)
+        assert isinstance(f1.returns, nodes.Const)
+        assert isinstance(f1.body[0], nodes.Pass)
+        assert (f1.lineno, f1.col_offset) == (1, 0)
+        assert (f1.end_lineno, f1.end_col_offset) == (6, 8)
+        assert (f1.returns.lineno, f1.returns.col_offset) == (5, 5)
+        assert (f1.returns.end_lineno, f1.returns.end_col_offset) == (5, 9)
+        assert (f1.body[0].lineno, f1.body[0].col_offset) == (6, 4)
+        assert (f1.body[0].end_lineno, f1.body[0].end_col_offset) == (6, 8)
+
+        # pos only arguments
+        # TODO fix column offset: arg -> arg (AssignName)
+        assert isinstance(f1.args.posonlyargs[0], nodes.AssignName)
+        assert (f1.args.posonlyargs[0].lineno, f1.args.posonlyargs[0].col_offset) == (2, 4)
+        assert (f1.args.posonlyargs[0].end_lineno, f1.args.posonlyargs[0].end_col_offset) == (2, 10)
+        assert isinstance(f1.args.posonlyargs_annotations[0], nodes.Name)
+        assert (
+            f1.args.posonlyargs_annotations[0].lineno, f1.args.posonlyargs_annotations[0].col_offset
+        ) == (2, 7)
+        assert (
+            f1.args.posonlyargs_annotations[0].end_lineno, f1.args.posonlyargs_annotations[0].end_col_offset
+        ) == (2, 10)
```

Comment:
```suggestion
        annotation = f1.args.posonlyargs_annotations[0]
        assert (annotation.lineno, annotation.col_offset) == (2, 7)
        assert (annotation.end_lineno, annotation.end_col_offset) == (2, 10)
```

There's very long line with long name that are duplicated, maybe doing that would help if we need to modify those tests later on ?

Reply:
Don't know about that. I think I prefer the current option as it doesn't hide the path. Otherwise you just end up introducing a lot of helper variables  which itself add complexity.

I explicitly chose to disable black in two functions to not have line breaks. Those made it even more difficult to read.